client/java/src/main/java/io/openlineage/client/utils/DatasetIdentifierUtils.java

```diff
@@ -5,88 +5,94 @@
 
-  /** Whether the current host is a Windows machine. */
-  public static final boolean WINDOWS = System.getProperty("os.name").startsWith("Windows");
+  /** The authority separator. */
+  private static final String AUTHORITY_SEPARATOR = "//";
 
-  /** Pre-org.apache.hadoop.shaded.com.iled regular expressions to detect path formats. */
-  private static final Pattern HAS_DRIVE_LETTER_SPECIFIER = Pattern.compile("^/?[a-zA-Z]:");
+  /** The scheme separator. */
+  private static final String SCHEME_SEPARATOR = ":";
 
   public static DatasetIdentifier fromURI(URI uri) {
     return fromURI(uri, DEFAULT_SCHEME);
   }
 
+  /* Build DatasetIdentifier from URI.
+   * Namespace is `file` or `scheme://authority`, name is the remaining path
+   */
+  @SneakyThrows
   public static DatasetIdentifier fromURI(URI uri, String defaultScheme) {
-    if (isAbsoluteAndSchemeAuthorityNull(uri)) {
-      return new DatasetIdentifier(uri.getPath(), defaultScheme);
-    }
+    URI namespace = new URI(uri.getScheme(), uri.getAuthority(), SEPARATOR, null, null);
+    String name = uri.getPath();
+    return fromComponents(namespace, name, defaultScheme);
+  }
 
-    String path = uri.getPath();
-    String name =
-        ("/".equals(path) || path.isEmpty())
-            ? "/"
-            : Optional.of(path)
-                .map(DatasetIdentifierUtils::removeLastSlash)
-                .map(DatasetIdentifierUtils::removeFirstSlashIfSingleSlashInString)
-                .get();
-
-    String namespace =
-        Optional.ofNullable(uri.getAuthority())
-            .map(a -> String.format("%s://%s", uri.getScheme(), a))
-            .orElseGet(() -> (uri.getScheme() != null) ? uri.getScheme() : defaultScheme);
-
-    return new DatasetIdentifier(name, namespace);
+  /* Build DatasetIdentifier from namespace (as URI) and name.
+   */
+  public static DatasetIdentifier fromComponents(URI namespace, String name) {
+    return fromComponents(namespace, name, DEFAULT_SCHEME);
   }
 
-  private static String removeFirstSlashIfSingleSlashInString(String name) {
-    if (name.chars().filter(x -> x == File.separatorChar).count() == 1 && name.startsWith("/")) {
-      return name.substring(1);
-    }
-    return name;
+  public static DatasetIdentifier fromComponents(URI namespace, String name, String defaultScheme) {
+    String normalizedNamespace = normalizeNamespace(namespace, defaultScheme);
+    String normalizedName = normalizeName(name);
+    return new DatasetIdentifier(normalizedName, normalizedNamespace);
   }
 
-  private static String removeLastSlash(String name) {
-    if (!name.isEmpty() && name.charAt(name.length() - 1) == File.separatorChar) {
-      return name.substring(0, name.length() - 1);
+  /* Convert namespace from URI to normalized string.
+   * Final slashes are removed, `/path` and `file:///path` are converted to `file:/path`
+   */
+  public static String normalizeNamespace(URI namespaceUri, String defaultScheme) {
```

Comment:
Any reason for making this method public ?

Reply:
It is required by #2767. But I can make it private in this PR, and convert to public in #2767.

Reply:
it's ok, no need to do that. thanks for answering.